flink-connector-aws/flink-connector-dynamodb/src/test/java/org/apache/flink/connector/dynamodb/source/enumerator/tracker/ShardGraphTrackerTest.java

```diff
@@ -0,0 +1,67 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.flink.connector.dynamodb.source.enumerator.tracker;
+
+import org.junit.jupiter.api.Test;
+import software.amazon.awssdk.services.dynamodb.model.Shard;
+
+import java.util.Arrays;
+import java.util.List;
+
+import static org.apache.flink.connector.dynamodb.source.util.TestUtil.generateShard;
+import static org.apache.flink.connector.dynamodb.source.util.TestUtil.generateShardId;
+import static org.assertj.core.api.Assertions.assertThat;
+
+/**
+ * Tests the {@link SplitGraphInconsistencyTracker} class to verify that it correctly discovers
+ * inconsistencies within the shard graph.
+ */
+public class ShardGraphTrackerTest {
+    @Test
+    public void testShardGraphTrackerHappyCase() {
+        List<Shard> shards =
+                Arrays.asList(
+                        // shards which don't have a parent
+                        generateShard(0, "1400", "1700", null),
+                        generateShard(1, "1500", "1800", null),
+                        // shards produced by rotation of parents
+                        generateShard(2, "1710", null, generateShardId(0)),
+                        generateShard(3, "1520", null, generateShardId(1)));
+        SplitGraphInconsistencyTracker shardGraphTracker = new SplitGraphInconsistencyTracker();
+        shardGraphTracker.addNodes(shards);
+        assertThat(shardGraphTracker.inconsistencyDetected()).isFalse();
+        assertThat(shardGraphTracker.getNodes()).containsExactlyInAnyOrderElementsOf(shards);
+    }
+
+    @Test
+    public void testShardGraphTrackerDetectsInconsistencies() {
+        List<Shard> shards =
+                Arrays.asList(
+                        // shards which don't have a parent
+                        generateShard(0, "1400", "1700", null),
+                        generateShard(1, "1500", "1800", null),
+                        // shards produced by rotation of parents
+                        generateShard(2, "1710", null, generateShardId(0)));
```

Comment:
Can we put in a later `closedLeafNode` so that we can test that the earlier one is detected? Best in a new test